#Python pipeline for generating protein files matching PAO1 or PAK, depending on the protein. Takes excel input files generated by NCBI-BLASTn and compares them to nucleotide FASTA files for the gene in question. Nucleotide sequences pulled out by that function are then translated. These files are manipulated using Pandas. Before translation, to ensure we don't have false truncations, we will delete all duplicates - aka any strains that are sequenced twice on different contigs, and may have been "truncated" onto two contigs. The protein sequence is then outputted into a txt file which will be used to generate dissimilarity graphs using R. All code in in UsefulFunctions (UF) python file.
import UsefulFunction_kao as UF
import pandas as pd
import argparse

parser = argparse.ArgumentParser(description = 'This program makes a MatPlotLib histogram showing length of amino acid sequences for a given gene in Pseudomonas aeruginosa')
parser.add_argument("Excel", type=str, help="First enter excel file")
args = parser.parse_args()

withtitle = UF.SplitExcel(args.Excel)

cleanedseqs = UF.Cleanup(withtitle)

allproteins = UF.Translate(cleanedseqs)

name = str(args.Excel)[:4]

df = pd.DataFrame(allproteins)

#export to CSV
df.to_csv(name + ".txt", sep='\t', index=False, header=False)